Feature/gpu support

agjharms opened the pull request:
# Feature/GPU Support

This pull request addresses issue #597 by adding a configuration option for the Node to allow the algorithm container access to the GPUs. This is done by creating and passing on a list of DeviceRequest objects from the docker-py package. Currently the only options are to block GPU access or allow access to all the GPUs on a machine, but there is an opportunity to make extend the configurability.

To interpret the Node configuration in DockerManager and then get the DeviceRequest to the algorithm container, I had to modify the arguments of DockerTaskManager. An alternative without this alteration would be to pass along the configuration through the algorithm environment variables, but seen as these configuration variables do not play a role _inside_ the container, I didn't think this was the best place to put this. 

It was not really clear to me which branch I should create the pull request for, dev3 seems to be somewhat behind main and I didn't want to make the choice for a release for you. If the PR should go somewhere else, please let me know.

frankcorneliusmartin commented on vantage6-node/vantage6/node/docker/docker_manager.py:
```diff
@@ -200,6 +206,21 @@ def _set_database(self, databases: Union[Dict, List]) -> None:
                                      'type': db_type}
         self.log.debug(f"Databases: {self.databases}")
 
+    def _set_algorithm_device_requests(self, device_requests_config: Dict):
```
* In python 3.10+ thypes can be simply `dict`.
* set return type to `None`

frankcorneliusmartin commented on vantage6-node/vantage6/node/docker/task_manager.py:
```diff
@@ -41,7 +41,8 @@ def __init__(self, image: str, vpn_manager: VPNManager, node_name: str,
                  databases: dict, docker_volume_name: str,
-                 alpine_image: Union[str, None] = None):
+                 alpine_image: Union[str, None] = None,
+                 device_requests: Union[List, None] = None):
```
Also something cool in python 3.10+, you can write `Union[str, None]` as `str | None` much more readable.

frankcorneliusmartin commented on vantage6-node/vantage6/node/docker/task_manager.py:
```diff
@@ -427,7 +438,7 @@ def _setup_environment_vars(self, algorithm_env: Dict = {},
                              f"{self.databases.keys()}. This is likely to "
                              "result in an algorithm crash.")
-            self.debug(f"User specified database: {database}")
+            self.log.debug(f"User specified database: {database}")
```
Thx

frankcorneliusmartin commented:
* I changed the merge branch to `release/3.9`
* Added the issue to the project board 
* @agjharms Could you resolve the merge conflict
* @agjharms Could you fix the minor comments (/ question)

Other than that I think this looks good. Development for `3.9` is complete from our end so I expect that we release it before the end of the month.

agjharms commented on vantage6-node/vantage6/node/docker/docker_manager.py:
```diff
@@ -200,6 +206,21 @@ def _set_database(self, databases: Union[Dict, List]) -> None:
+        """
+        device_requests = []
+        if device_requests_config.get('gpu', False):
+            device_requests.append(docker.types.DeviceRequest(count=-1, capabilities=[['gpu']]))
```
Yes, with '-1' access is given to all GPUs. Currently I think that Vantage6 nodes are often run in a separate (virtual) machine, so it's not a big deal to allot all of the resources to the node. I can imagine a situation in the future where a user has multiple Vantage6 nodes running on a machine with multiple GPUs, and wishes to designate specific GPUs to each node. In that case you could use the `device_ids` argument for DeviceRequest to pick out specific GPUs. I didn't have the opportunity to look into making this configurable in a nice way, so I reduced it to a boolean option for now with the option to extend it if the need arises.

agjharms commented:
Thanks for the feedback. I have addressed the points that you mentioned. I tried to rebase the `release/3.9` branch on the feature branch I had in my fork to accommodate the change in target branches, but that seems to not have been the right approach. I will fix that.

agjharms commented on vantage6-node/vantage6/node/docker/task_manager.py:
```diff
@@ -101,6 +105,12 @@ def __init__(self, image: str, vpn_manager: VPNManager, node_name: str,
 
+        # set device requests
+        self.device_requests = []
+        if device_requests is not None:
```
You're right that `if device_requests` is a bit cleaner. Thanks for the heads-up on the `is` operator.
